File: clients/javascript/debug-interface.js
const ClientMessage = require('./codegame/client-message');
const DebugState = require('./debugging/debug-state');
const DebugCommand = require('./debugging/debug-command');
const DebugData = require('./debugging/debug-data');

class DebugInterface {
    constructor(stream) {
        this.stream = stream;
    }

    async addCircle(pos, radius, color) {
        await this.add(new DebugData.Circle(pos, radius, color));
    }
    
    async addLine(point1, point2, width, color) {
        await this.add(new DebugData.Line(point1, point2, width, color));
    }
    
    async addRect(corner1, corner2, color) {
        await this.add(new DebugData.Rect(corner1, corner2, color));
    }
    
    async addText(text, pos, size, align, color) {
        await this.add(new DebugData.Text(text, pos, size, align, color));
    }

    async add(debugData) {
        await this.send(new DebugCommand.Add(debugData));
    }
    
    async clear() {
        await this.send(new DebugCommand.Clear());
    }
    
    async setAutoFlush(enable) {
        await this.send(new DebugCommand.SetAutoFlush(enable));
    }
    
    async flush() {
        await this.send(new DebugCommand.Flush());
    }

    async send(command) {
        await (new ClientMessage.DebugMessage(command)).writeTo(this.stream);
        await this.stream.flush();
    }

    async getState() {
        await (new ClientMessage.RequestDebugState()).writeTo(this.stream);
        await this.stream.flush();
        return await DebugState.readFrom(this.stream);
    }
}

module.exports = DebugInterface;

File: clients/ruby/debug_interface.rb
class DebugInterface
    def initialize(stream)
        @stream = stream
    end

    def add_circle(pos, radius, color)
        add(Debugging::DebugData::Circle.new(pos, radius, color))
    end
    
    def add_line(point1, point2, width, color)
        add(Debugging::DebugData::Line.new(point1, point2, width, color))
    end
    
    def add_rect(corner1, corner2, color)
        add(Debugging::DebugData::Rect.new(corner1, corner2, color))
    end
    
    def add_text(text, pos, size, align, color)
        add(Debugging::DebugData::Text.new(text, pos, size, align, color))
    end

    def add(debug_data)
        send(Debugging::DebugCommand::Add.new(debug_data))
    end
    
    def clear()
        send(Debugging::DebugCommand::Clear.new())
    end
    
    def set_auto_flush(enable)
        send(Debugging::DebugCommand::SetAutoFlush.new(enable))
    end
    
    def flush()
        send(Debugging::DebugCommand::Flush.new())
    end

    def send(command)
        Codegame::ClientMessage::DebugMessage.new(command).write_to(@stream)
        @stream.flush()
    end

    def get_state()
        Codegame::ClientMessage::RequestDebugState.new().write_to(@stream)
        @stream.flush()
        return Debugging::DebugState.read_from(@stream)
    end
end